mirrord/agent/src/main.rs

```diff
@@ -390,67 +444,71 @@ async fn start_agent() -> Result<()> {
-                            match ClientConnectionHandler::new(
-                                    client_id,
-                                    stream,
-                                    pid,
-                                    args.ephemeral_container,
-                                    sniffer_command_tx,
-                                    stealer_command_tx,
-                                    dns_sender,
-                                )
-                                .and_then(|client| client.start(cancellation_token))
-                                .await
-                                {
-                                    Ok(_) => {
-                                        trace!(
-                                            "ClientConnectionHandler::start -> Client {} disconnected",
-                                            client_id
-                                        );
-                                    }
-                                    Err(e) => {
-                                        error!(
-                                            "ClientConnectionHandler::start -> Client {} disconnected with error: {}",
-                                            client_id, e
-                                        );
-                                    }
-                                }
-                            client_id
-
-                        });
-                        clients.push(client);
-                    },
-                    Err(AgentError::ConnectionLimitReached) => {
-                        error!("start_client -> Ran out of connections, dropping new connection");
-                    },
-                    // Propagate all errors that are not ConnectionLimitReached.
-                    err => { err?; },
-                }
+                if let Some(client) = state.new_connection(
+                    stream,
+                    sniffer_command_tx.clone(),
+                    stealer_command_tx.clone(),
+                    cancellation_token.clone(),
+                    dns_sender.clone(),
+                    args.ephemeral_container,
+                    pid
+                ).await? {clients.push(client) };
             },
-            Some(client) = clients.next() => {
-                let client_id = client?;
-                state.remove_client(client_id).await?;
-            },
-            _ = tokio::time::sleep(std::time::Duration::from_secs(args.communication_timeout.into())) => {
-                if state.no_clients_left() {
-                    trace!("Main thread timeout, no clients connected.");
-                    break;
+            // Logic here is if we have already accepted a client, then we drop immediately when all clients disconnect.
+            // if not, we wait for the communication timeout to expire before dropping. (to have more time for the first
+            // client to connect)
```

Comment:
> we wait for the communication timeout to expire before dropping

Where is that?
> (to have more time for the first client to connect)

Didn't the first client already connect (above the loop) if we reached here?

Reply:
outdated comment, removed